source/npm/util.js

```diff
@@ -107,3 +109,74 @@ exports.checkIgnoreStrategy = ({files}) => {
 		`);
 	}
 };
+
+// New files added part of .npmignore or not part of files-attribute in "package.json"
+exports.checkNewFiles = async (newFiles, filesFromPackageDotJson) => {
+	const rootDir = pkgDir.sync();
+	const npmignoreExists = fs.existsSync(path.resolve(rootDir, '.npmignore'));
+
+	let result = [];
+
+	if (filesFromPackageDotJson) {
+		result = getFilesNotPartOfFileAttribute(filesFromPackageDotJson, newFiles);
+	}
+
+	if (npmignoreExists) {
+		result = result.concat(await getFilesIgnoredByDotnpmignore(newFiles));
+	}
+
+	return result;
+};
+
+async function getFilesIgnoredByDotnpmignore(fileList) {
+	if (!Array.isArray(fileList)) {
+		throw new TypeError('expected array, but got {typeof fileList}');
+	}
+
+	const result = [];
+	await ignoreWalker({
+		path: pkgDir.sync(),
+		ignoreFiles: ['.npmignore']
+	})
+		.then(WhiteList => {
+			if (!Array.isArray(WhiteList)) {
+				throw new TypeError(`Expected array, but got ${typeof WhiteList} `);
+			}
```

Comment:
What is this check for exactly? `ignoreWalker` is something `np` calls and the error is something that the user sees.

If `ignoreWalker` _really_ might return something that's not an array (but [it doesn't look like it](https://github.com/npm/ignore-walk/blob/0e4f87adccb3e16f526d2e960ed04bdc77fd6cca/index.js#L34-L37)) then `np` should account for this case specifically rather than throwing a generic `TypeError`.

---

This probably applies to other typechecks as well: don't type-check arguments unless necessary or provide useful errors otherwise.

Reply:
The error should really be thrown when ```ignoreWalker``` returns something different to an array. With my latest commit i removed this check, because a array was always returned.